lib/edgehog_device_forwarder/connections/token_data.ex

```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Connections.TokenData do
+  defstruct realm: nil, device_id: nil, socket: nil
+
+  @type t :: %__MODULE__{}
+  @type device_id :: String.t()
+  @type realm :: String.t()
+  @type socket :: pid()
+
+  @spec new(socket) :: t
+  def new(socket), do: %__MODULE__{socket: socket}
```

Comment:
If you added this function to ensure that `TokenData` always gets created with a socket, note that a user can always construct the struct manually. Adding `@enforce_keys` should help you with that, consider using [`TypedStruct`](https://hexdocs.pm/typed_struct_uberbrodt/TypedStruct.html), which we're recently introducing in our codebases since it reduces some of the noise associated with (well-typed) struct definitions.

Also, since you're using the `new` function to guarantee a valid initialization, you should guard on `socket` being a `pid`.

Reply:
alright I'll look into adding TypedStructs to the project